File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.cpp
#include <iostream>
#include <vector>
#include <bitset>

using namespace std;

long dfs(int idx, long visited, const vector<int> &cards) {
    if (visited & (1 << idx)) return visited;
    visited |= (1 << idx);
    int next = cards[idx] - 1;
    return dfs(next, visited, cards);
}

int solution(vector<int> cards) {
    size_t n = cards.size();
    int answer = 0;
    for (int i = 0; i < n; i++) {
        long visited = dfs(i, 0, cards);
        int a = __builtin_popcount(visited);
        for (int j = 0; j < n; j++) {
            long visited2 = dfs(j, visited, cards);
            int b = __builtin_popcount(visited2) - a;
            answer = max(answer, a * b);
        }
    }
    return answer;
}

File: 프로그래머스/2/131130. 혼자 놀기의 달인/혼자 놀기의 달인.py
def solution(cards):
    def dfs(cur, visited):
        if visited & (1 << cur): return visited
        visited |= (1 << cur)
        next = cards[cur] - 1
        return dfs(next, visited)
    
    def bit_count(x):
        return bin(x).count('1')
    
    answer = 0
    for i in range(len(cards)):
        visited = dfs(i, 0)
        a = bit_count(visited)
        for j in range(len(cards)):
            visited2 = dfs(j, visited)
            b = bit_count(visited2 - visited)
            answer = max(answer, a * b)
    return answer